src/thor/bidirectional_astar.cc

```diff
@@ -638,6 +684,24 @@ BidirectionalAStar::GetBestPath(valhalla::Location& origin,
         continue;
       }
 
+      // Check if this branch can be pruned. It's implementation of the reach-based pruning technique
+      // for bidirectional astar: https://repub.eur.nl/pub/16100/ei2009-10.pdf .
+      if (cost_threshold_ != std::numeric_limits<float>::max() &&
+          fwd_pred.predecessor() != kInvalidLabel) {
+        const auto& fwd_pred_pred = edgelabels_forward_[fwd_pred.predecessor()];
+        if (auto pred_tile = graphreader.GetGraphTile(fwd_pred_pred.endnode());
+            pred_tile != nullptr) {
+          // Estimate lower bound cost for the shortest path that goes through the current edge.
+          float route_lower_bound =
+              fwd_pred_pred.cost().cost + fwd_pred.transition_cost().cost + rev_pred.sortcost() -
+              astarheuristic_reverse_.Get(pred_tile->get_node_ll(fwd_pred_pred.endnode()));
```

Comment:
Estimate lower bound cost for the route that goes through the current edge. The technique from the article was adopted to the edges (in the article used vertices). Let me describe the formula for vertices. Let's suppose the current vertex we pulled out of the forward queue is `v` (`v'` - a vertex with min sortcost in the reverse queue), so the are following equations are true:
```
cost_f(v) + sortcost_r(v') - h_r(v) <= cost_f(v) + sortcost_r(v) - h_r(v) = cost_f(v) + cost_r(v) + h_r(v) - h_r(v) =
cost_f(v)  + cost_r(v) = cost(shortest path through v)

where
cost, sortcost, h - costing function, sortcost, heuristic function;
*_f, *_r - indicates forward and reverse directions respectively.
```
these equations say that `cost_f(v) + sortcost_r(v') - h_r(v) ` is a lower bound cost for the route through `v`. Using this logic we can replace vertices by edges and get approximately the same formula for our case.